File: h6_sazonalidade_dia/h6_sazonalidade.R
# Carregar pacotes necessários
library(dplyr)
library(readr)
library(lubridate)
library(stats)

# Carregar os dados
df_carrinhos <- read_csv("_data/clean/e_commerce_carrinhos.csv")
df_clientes <- read_csv("_data/clean/e_commerce_clientes.csv")
df_pedidos <- read_csv("_data/clean/e_commerce_pedidos.csv")

# Análise de tipos
str(df_pedidos)

# Ver colunas
names(df_pedidos)

head(df_pedidos)

# Convertendo data - string para datetime
df_pedidos$data <- dmy_hms(df_pedidos$data)

# Obter apenas a data, sem o horário
df_pedidos$data_aux <- as.Date(df_pedidos$data)

# Agrupar por ID do pedido e obter a primeira ocorrência de cada data para cada pedido
df_pedidos_unicos <- df_pedidos %>% 
  group_by(id) %>% 
  slice(1) %>%
  ungroup()

# Contar pedidos por data
contagem_por_data <- df_pedidos_unicos %>% 
  group_by(data_aux) %>% 
  summarise(contagem = n()) %>% 
  arrange(desc(contagem))

print(contagem_por_data)

# Criando nova coluna - dia da semana
df_pedidos$dia_semana <- weekdays(df_pedidos$data)

head(df_pedidos)

# Agrupar por ID do pedido e obter a primeira ocorrência de cada data para cada pedido
df_pedidos_unicos <- df_pedidos %>% 
  group_by(id) %>% 
  slice(1) %>%
  ungroup()

# Agrupar por 'dia_semana' e contar o número de pedidos únicos
contagem_por_dia <- df_pedidos_unicos %>% 
  group_by(dia_semana) %>% 
  summarise(contagem = n()) %>% 
  arrange(desc(contagem))

print(contagem_por_dia)

# Frequência esperada se os pedidos fossem uniformemente distribuídos
frequencia_esperada <- nrow(df_pedidos_unicos) / 7
print(frequencia_esperada)

# Extrair a lista de contagens observadas e a lista de dias da semana
contagens_observadas <- contagem_por_dia$contagem
dias_da_semana <- contagem_por_dia$dia_semana

# Calcular a frequência esperada (assumindo uma distribuição uniforme)
frequencia_esperada <- sum(contagens_observadas) / length(dias_da_semana)
frequencias_esperadas <- rep(frequencia_esperada, length(dias_da_semana))

# Realizar o teste qui-quadrado
chi2_test <- chisq.test(contagens_observadas, p = frequencias_esperadas / sum(frequencias_esperadas))

# Exibir os resultados
print(paste("Estatística qui-quadrado:", chi2_test$statistic))
print(paste("Valor-p:", chi2_test$p.value))

if (chi2_test$p.value < 0.05) {
  print("Rejeitamos a hipótese nula: há evidência de sazonalidade nos dias da semana.")
} else {
  print("Não rejeitamos a hipótese nula: não há evidência de sazonalidade nos dias da semana.")
}

File: h6_sazonalidade_dia/h6_sazonalidade.py
import pandas as pd
import scipy.stats as stats

#criando os dfs
df_carrinhos = pd.read_csv("_data\clean\e_commerce_carrinhos.csv")
df_clientes = pd.read_csv("_data\clean\e_commerce_clientes.csv")
df_pedidos = pd.read_csv("_data\clean\e_commerce_pedidos.csv")

#analise de tipos
print(df_pedidos.dtypes)

#vendo colunas
print(df_pedidos.columns)

#Listando valores únicos nas colunas
for coluna in df_pedidos.columns:
    valores_unicos = df_pedidos[coluna].unique()
    print(f'Coluna "{coluna}" - Valores únicos: {valores_unicos}')

#convertendo data - str > datetime
df_pedidos['data'] = pd.to_datetime(df_pedidos['data'], format='%d/%m/%Y %H:%M:%S')

#contando pedidos por data
contagem_por_data = df_pedidos.groupby('data').size().reset_index(name='contagem')
print(contagem_por_data.sort_values('contagem', ascending= False))

#criando nova coluna - dia da semana
df_pedidos['dia_semana'] = df_pedidos['data'].dt.day_name()

print(df_pedidos)

#contando pedidos por dia
contagem_por_dia = df_pedidos.groupby('dia_semana').size().reset_index(name='contagem')
print(contagem_por_dia.sort_values('contagem', ascending= False))

# Contagem dos pedidos por dia da semana
contagem_dias = df_pedidos['dia_semana'].value_counts()

print(contagem_dias)

# Frequência esperada se os pedidos fossem uniformemente distribuídos
frequencia_esperada = len(df_pedidos) / 7

print(frequencia_esperada)

# Realizar o teste qui-quadrado
chi2, p_valor = stats.chisquare(contagem_dias, [frequencia_esperada]*7)

# Exibir os resultados
print(f"Estatística qui-quadrado: {chi2}")
print(f"Valor-p: {p_valor}")

if p_valor < 0.05:
    print("Rejeitamos a hipótese nula: há evidência de sazonalidade nos dias da semana.")
else:
    print("Não rejeitamos a hipótese nula: não há evidência de sazonalidade nos dias da semana.")